libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import { EnvironmentVariableNames } from '../../api/EnvironmentConfiguration';
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import { NullOperationRunner } from './NullOperationRunner';
+import { Operation } from './Operation';
+import { OperationStatus } from './OperationStatus';
+import { ShellOperationRunner } from './ShellOperationRunner';
+import {
+  formatCommand,
+  getCustomParameterValuesByPhase,
+  getDisplayName,
+  getScriptToRun
+} from './ShellOperationRunnerPlugin';
+
+export const PLUGIN_NAME: 'ShardedPhasedOperationPlugin' = 'ShardedPhasedOperationPlugin';
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStrings = {
+  SHARD_INDEX: '{shardIndex}',
+  SHARD_COUNT: '{shardCount}'
+} as const;
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStringRegex = {
+  SHARD_INDEX: new RegExp(TemplateStrings.SHARD_INDEX, 'g'),
+  SHARD_COUNT: new RegExp(TemplateStrings.SHARD_COUNT, 'g')
+} as const;
+
+/**
+ * Phased command that shards a phase into multiple operations.
+ */
+export class ShardedPhasedOperationPlugin implements IPhasedCommandPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tap(PLUGIN_NAME, spliceShards);
+  }
+}
+
+function spliceShards(existingOperations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
+  const { rushConfiguration } = context;
+
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project, settings: operationSettings } = operation;
+    if (phase && project && operationSettings?.sharding && !operation.runner) {
+      const { count: shards, shardScriptConfiguration } = operationSettings.sharding;
+
+      existingOperations.delete(operation);
+
+      const collatorNode: Operation = new Operation({
+        phase,
+        project,
+        settings: operationSettings
+      });
+      existingOperations.add(collatorNode);
+
+      const parentFolderFormat: string =
+        operationSettings.sharding.outputFolderArgument?.parentFolderName ??
+        `.rush/operations/${phase.logFilenameIdentifier}/shards`;
+
+      const parentFolder: string = parentFolderFormat;
+
+      const collatorDisplayName: string = `${getDisplayName(phase, project)} - collate`;
+      const collatorRawScript: string | undefined = getScriptToRun(project, phase.name, undefined);
+
+      if (collatorRawScript === undefined && phase.missingScriptBehavior === 'error') {
```

Comment:
Handle the `phase.missingScriptBehavior === 'log'` case?

Reply:
Ah I see line 99. Is that what `'log'` does for non-sharded operations?

Reply:
I believe so, I adjusted this to a single method in `ShellOperationRunnerPlugin` that handles this.